Feature/product detail v1

nbradbury opened the pull request:
This PR represents the first take on product detail. There are a number of things missing (see #913 for details), but all product properties not listed separately in #913 should be supported in this PR.

Because this is already quite large, I may opt to tackle issues found in this PR separately as well. We can hold off on a design review until the design has been solidified and the master feature branch is ready.

**Note:** [This FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1174) and [this FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1184) should be reviewed and merged first.

![Screenshot_1553261155](https://user-images.githubusercontent.com/3903757/54825905-fe753900-4c84-11e9-887f-8f5951e68f14.png)

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:
```diff
@@ -420,6 +421,13 @@ class MainActivity : AppCompatActivity(),
 
+    override fun showProductDetail(remoteProductId: Long) {
+        showBottomNav()
+        // TODO analytics
```
Left a TODO for analytics...should this be added?

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderDetailFragment.kt:
```diff
@@ -185,6 +186,7 @@ class OrderDetailFragment : Fragment(), OrderDetailContract.View, OrderDetailNot
                     productImageMap,
                     false,
                     currencyFormatter.buildFormatter(order.currency),
+                    this,
```
np: To make it clearer what `this` is, I recommend using named arguments here `orderListener = this, productListener = this`

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderFulfillmentFragment.kt:
```diff
@@ -83,7 +84,14 @@ class OrderFulfillmentFragment : Fragment(), OrderFulfillmentContract.View, View
+                productImageMap,
+                true,
+                currencyFormatter.buildFormatter(order.currency),
+                this,
```
np: Same as before, recommend named arguments here for clarity

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderFulfillmentFragment.kt:
```diff
@@ -124,4 +132,20 @@ class OrderFulfillmentFragment : Fragment(), OrderFulfillmentContract.View, View
+
+    override fun openOrderProductList(order: WCOrderModel) {
+        TODO("not implemented")
+    }
```
Since neither of these need to be implemented, I recommend removing `OrderActionListener` from `OrderFulfillmentContract`

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/OrderProductListFragment.kt:
```diff
@@ -80,6 +81,21 @@ class OrderProductListFragment : Fragment(), OrderProductListContract.View {
+                productImageMap,
+                true,
+                currencyFormatter.buildFormatter(order.currency),
+                null,
```
np: named arguments recommended here as well

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailFragment.kt:
```diff
@@ -0,0 +1,418 @@
+
+        product.getFirstImageUrl()?.let {
+            val imageWidth = DisplayUtils.getDisplayPixelWidth(activity)
+            val imageHeight = resources.getDimensionPixelSize(R.dimen.product_detail_image_height)
```
The product image is not the full image, but rather a zoomed in, cut off version of itself. I know there is another PR that further tweaks this image portion so I'll bring it up there, but thought I would mention it here as well.

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailFragment.kt:
```diff
@@ -0,0 +1,418 @@
+        addPurchaseDetailsCard(product)
+    }
+
+    private fun addPrimaryCard(product: WCProductModel) {
```
I really love these reusable cards...so slick! 🎆

AmandaRiu commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailFragment.kt:
```diff
@@ -0,0 +1,418 @@
+        if (product.ratingCount > 0) {
+            addPropertyView(
+                    DetailCard.Primary,
+                    R.string.product_reviews,
```
I know this kinda breaks the current ease of use here in dynamic fields, but I recommend using `FlowLayout` for the review stars so if space runs low, the view will automatically change to a vertically aligned title/value view. Doesn't need to be in this PR, just noting it here.

nbradbury commented:
Thanks for the review, @AmandaRiu! I added named arguments in e37face, and in 3e01d3d I removed the order listener from the fulfillment fragment.

I updated #913 to include the `FlowLayout` idea. I'm hoping that won't be necessary, though.

Regarding the image cropping, I wasn't sure of the best way to handle it, but as with media in WPAndroid I thought it best to keep the image a consistent height and crop it.

One alternative is to keep the image a fixed height without cropping and use a solid background to fill the space, which works but isn't as attractive as cropping:

![Screenshot_1553944543](https://user-images.githubusercontent.com/3903757/55275723-8a5c1600-52c0-11e9-8fe8-1688dce42f3c.png)

Another alternative is to use a fixed width and let the image determine the height, but this doesn't work well with images that are taller than they are wide:

![Screenshot_1553944830](https://user-images.githubusercontent.com/3903757/55275726-8fb96080-52c0-11e9-9229-deaa32b85fe7.png)

My vote would be to keep cropping the image, but enable tapping on the image to see it fullscreen without cropping.

AmandaRiu commented:
> My vote would be to keep cropping the image, but enable tapping on the image to see it fullscreen without cropping. I think we can punt on it for now and address it when we're ready for design review

I like that option the best!!

nbradbury commented:
> ...the title bar changes from showing the name of the product to the order list title

Good catch @AmandaRiu! Looks like this only happens when "Don't keep activities" is enabled, and it turns out it's already fixed in #927 (where the product detail fragment is converted into an activity).
